src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNWeight.java

```diff
@@ -74,57 +74,66 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
+            List<String> engineFiles = reader.getSegmentInfo().files().stream()
+                    .filter(fileName -> fileName.endsWith(suffix))
                     .collect(Collectors.toList());
 
-            if(hnswFiles.isEmpty()) {
-                logger.debug("[KNN] No hnsw index found for field {} for segment {}",
+            if(engineFiles.isEmpty()) {
+                logger.debug("[KNN] No engine index found for field {} for segment {}",
                         knnQuery.getField(), reader.getSegmentName());
                 return null;
             }
 
             FieldInfo queryFieldInfo = reader.getFieldInfos().fieldInfo(knnQuery.getField());
             Map<String, String> fieldAttributes = queryFieldInfo.attributes();
 
-            /**
-             * in the name of the file. As of now we have one version 2.0.11 with NMSLIB and 1.6.5 with FAISS
-             * So deferring this to future releases
-             */
-
-            Path indexPath = PathUtils.get(directory, hnswFiles.get(0));
+            Path indexPath = PathUtils.get(directory, engineFiles.get(0));
             final KNNQueryResult[] results;
+            final KNNIndex index = knnIndexCache.getIndex(indexPath.toString(), knnQuery.getIndexName(), spaceType);
 
-            final KNNIndex index = knnIndexCache.getIndex(indexPath.toString(), knnQuery.getIndexName());
-            //we check instance of KNNINdex
-            if (fieldAttributes.containsValue(KNNEngine.NMSLIB.getKnnEngineName())) {
-                assert(index instanceof KNNNmsLibIndex);
-                results = index.queryIndex(
+            if ((fieldAttributes.containsValue(KNNEngine.NMSLIB.getName()) && index instanceof KNNNmsLibIndex)
+                    || (fieldAttributes.containsValue(KNNEngine.FAISS.getName())
+                    && index instanceof KNNFaissIndex)) {
+                results = index.query(
                         knnQuery.getQueryVector(),
                         knnQuery.getK()
                 );
             } else {
-                assert(index instanceof KNNFaissIndex);
-                results = index.queryIndex(
-                        knnQuery.getQueryVector(),
-                        knnQuery.getK()
-                );
+                throw new IllegalStateException("Unable to retrieve k-NN engine for index path: "
+                        + indexPath.toString());
             }
 
-            /**
+            /*
              * Scores represent the distance of the documents with respect to given query vector.
              * Lesser the score, the closer the document is to the query vector.
              * Since by default results are retrieved in the descending order of scores, to get the nearest
              * neighbors we are inverting the scores.
              */
+            if (results.length == 0) {
+                logger.debug("[KNN] Query yielded 0 results");
+                return null;
+            } else {
```

Comment:
else is redundant

Reply:
Right, I dont need that debug statement anyway.